### PR TITLE
Who has has what html reprs v2

### PR DESCRIPTION
**Pull request by jacobtomlinson**
Replacement for #4853 which was reverted with support for async calls.

cc @mrocklin @jrbourbeau @jakirkham @quasiben @crusaderky

### CODE REVIEW

File: distributed/tests/test_client.py [hunk 1]
```diff
@@ -3661,7 +3661,6 @@ async def test_async_whowhat(c, s, a, b):
     assert has_what == {a.address: (x.key,), b.address: ()}
 
 
-@pytest.mark.xfail(reason="Want to fix to use `WhoHas` + `WhatHas`")
 def test_client_repr(c):
```
Comment (jrbourbeau) [inline distributed/tests/test_client.py:3664]: Can we add a similar test which uses an `asynchronous=True` client?

**Review by jrbourbeau — COMMENTED** [review]
Thanks @jacobtomlinson! This looks great

@jakirkham @dantegd FYI this shouldn't cause any `cuml` breakages like we [were seeing earlier](https://github.com/dask/distributed/pull/4863#issuecomment-850544646) -- though do let us know if anything pops up

File: distributed/tests/test_client.py (hunk 1 above)
Comment (jacobtomlinson) [reply distributed/tests/test_client.py]: Done!

**Review by jrbourbeau — APPROVED** [review]
Thanks @jacobtomlinson! Just bumped CI to make sure test failures are unrelated to the changes here. Will merge after CI finishes

**Comment by jacobtomlinson** [comment]
@jrbourbeau failure looks unrelated, what do you think?

**Comment by jrbourbeau** [comment]
Yep, definitely unrelated (xref https://github.com/dask/distributed/issues/4862) -- merging

**Pull request merged**